### PR TITLE
[SYCLomatic] Support more configs for dpct::blas_gemm::experimental::matmul

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 1]
```diff
@@ -430,6 +448,13 @@ inline void int2float(queue_ptr q_ptr, void *int_ptr, bool is_host_ptr,
     });
   }
 }
+
+inline void multiply(queue_ptr q_ptr, float *result, float *a, float *b) {
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: This function should be a template function.
We also need a wrapper to do type dispatch.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 2]
```diff
@@ -484,62 +510,74 @@ inline sycl::event matmul(descriptor_ptr handle, matmul_desc_ptr compute_desc,
...
-        "dpct::blas_gemm::experimental::matmul() only supports data type "
-        "combinataions:\n  scale_type==int32 && a_type==int8 && b_type==int8 "
-        "&& c_type==int32,\n  scale_type==float && a_type==int8 && "
-        "b_type==int8 && c_type==int8 or\n  scale_type==float && a_type==int8 "
-        "&& b_type==int8 && c_type==int32.");
-  }
+      beta_is_zero = false;
+    // throw std::runtime_error("dpct::blas_gemm::experimental::matmul() does "
+    //                          "not support non-zero beta currently.");
+    // TODO: throw exception when beta is not 1
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: // TODO: throw exception when beta is not 1

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 3]
```diff
@@ -484,62 +510,74 @@ inline sycl::event matmul(descriptor_ptr handle, matmul_desc_ptr compute_desc,
...
+  }
+
+  // if (compute_desc->_epilogue != epilogue_t::nop) {
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]:
```suggestion
  if (compute_desc->_epilogue != epilogue_t::nop && compute_desc->_epilogue != epilogue_t::relu) {
```

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 4]
```diff
@@ -484,62 +510,74 @@ inline sycl::event matmul(descriptor_ptr handle, matmul_desc_ptr compute_desc,
...
+  // if (compute_desc->_trans_a != oneapi::mkl::transpose::nontrans) {
+  //   throw std::runtime_error("dpct::blas_gemm::experimental::matmul() only "
+  //                            "supports non-transposed matrix A currently.");
+  // }
+  // if (compute_desc->_trans_b != oneapi::mkl::transpose::trans) {
+  //   throw std::runtime_error("dpct::blas_gemm::experimental::matmul() only "
+  //                            "supports transposed matrix B currently.");
+  // }
+
+  // if (!(compute_desc->_scale_type == library_data_t::real_int32 &&
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: Update the rescriction.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 5]
```diff
@@ -551,50 +589,88 @@ inline sycl::event matmul(descriptor_ptr handle, matmul_desc_ptr compute_desc,
...
       sycl::event e = detail::matrix_transform<int>(
           q_ptr, b_desc->_rows, b_desc->_cols, b_desc->_ld, b_desc->_order,
           (const int *)b, new_ldb, order_t::col, (int *)new_b, {});
       transform_events.push_back(e);
     }
   }
+
+  if (!beta_is_zero && c_desc->_order != order_t::col) {
+    new_ldc = c_desc->_rows;
+    if (c_desc->_type == library_data_t::real_int8) {
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: Support more types for A/B/C/D.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 6]
```diff
@@ -676,48 +763,115 @@ inline sycl::event matmul(descriptor_ptr handle, matmul_desc_ptr compute_desc,
       dpct::dpct_memcpy(scales_alpha->get_data_handle(), alpha, sizeof(float),
                         automatic, *q_ptr);
     }
+    // multiply scale_a and scale_b
+    detail::multiply(q_ptr, (float *)(scales_alpha->get_data_handle()),
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: Need call a wrapper for type dispatch

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 7]
```diff
@@ -676,48 +763,115 @@ inline sycl::event matmul(descriptor_ptr handle, matmul_desc_ptr compute_desc,
...
+          sycl::reduction((float *)(amax_ptr), sycl::maximum<>());
+      cgh.depends_on({matmul_prim_event, scale_d_with_alpha_event});
+      cgh.parallel_for<dpct_kernel_name<class max_reduction>>(
+          sycl::range<2>(ld, cols), max_reduction,
+          [=](sycl::id<2> idx, auto &max) {
+            size_t row_idx = idx.get(0);
+            size_t col_idx = idx.get(1);
+            if (row_idx < rows) {
+              size_t linear_idx = row_idx + ld * col_idx;
+              max.combine(((float *)new_d)[linear_idx]);
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: Need call a wrapper for type dispatch

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 8]
```diff
@@ -676,48 +763,115 @@ inline sycl::event matmul(descriptor_ptr handle, matmul_desc_ptr compute_desc,
...
+    scale_d_event = q_ptr->submit([&](sycl::handler &cgh) {
+      cgh.depends_on(
+          {matmul_prim_event, scale_d_with_alpha_event, amax_d_event});
+      cgh.parallel_for<dpct_kernel_name<class scale_d>>(
+          sycl::range<2>(ld, cols), [=](sycl::id<2> idx) {
+            size_t row_idx = idx.get(0);
+            size_t col_idx = idx.get(1);
+            if (row_idx < rows) {
+              size_t linear_idx = row_idx + ld * col_idx;
+              ((float *)new_d)[linear_idx] =
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: Need call a wrapper for type dispatch

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 9]
```diff
@@ -676,48 +763,115 @@ inline sycl::event matmul(descriptor_ptr handle, matmul_desc_ptr compute_desc,
...
+  }
 
   sycl::event transform_d_event;
   if (d_desc->_order != order_t::col) {
     if (d_desc->_type == library_data_t::real_int8) {
       transform_d_event = detail::matrix_transform<std::int8_t>(
           q_ptr, d_desc->_rows, d_desc->_cols, new_ldd, order_t::col,
           (const std::int8_t *)new_d, d_desc->_ld, d_desc->_order,
-          (std::int8_t *)d, {scale_d_event, matmul_prim_event});
+          (std::int8_t *)d,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: Need support more types

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 10]
```diff
@@ -430,14 +463,93 @@ inline void int2float(queue_ptr q_ptr, void *int_ptr, bool is_host_ptr,
     });
   }
 }
+
+template <typename T> struct multiply_impl {
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: Add buffer version

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 11]
```diff
@@ -430,14 +463,93 @@ inline void int2float(queue_ptr q_ptr, void *int_ptr, bool is_host_ptr,
...
+          [=]() { result_T[0] = result_T[0] * a_T[0] * b_T[0]; });
+    });
+  }
+};
 #endif
+
+template <typename T> struct get_beta_value_impl {
+  int operator()(const void *beta, dpct::queue_ptr q_ptr) {
+    T beta_host;
+    q_ptr->memcpy(&beta_host, beta, sizeof(T)).wait();
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: Use dpct_memcpy

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 12]
```diff
@@ -430,14 +463,93 @@ inline void int2float(queue_ptr q_ptr, void *int_ptr, bool is_host_ptr,
...
+  }
+};
+
+template <typename T> struct scale_d_impl {
+  sycl::event operator()(const void *d_scale_ptr, void *d, size_t ld,
+                         size_t rows, size_t cols, dpct::queue_ptr q_ptr,
+                         dpct::library_data_t scale_type,
+                         std::vector<sycl::event> deps) {
+    auto new_d = (T *)d;
+    return q_ptr->submit([&](sycl::handler &cgh) {
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp:586]: Add buffer version

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 13]
```diff
@@ -430,14 +463,93 @@ inline void int2float(queue_ptr q_ptr, void *int_ptr, bool is_host_ptr,
...
+          });
+    });
+  }
+};
+
+template <typename T> struct amax_impl {
+  sycl::event operator()(void *amax_ptr, const void *new_d, size_t ld,
+                         size_t rows, size_t cols, dpct::queue_ptr q_ptr,
+                         std::vector<sycl::event> deps) {
+    return q_ptr->submit([&](sycl::handler &cgh) {
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp:638]: Add buffer version

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 14]
```diff
@@ -673,51 +913,112 @@ inline sycl::event matmul(descriptor_ptr handle, matmul_desc_ptr compute_desc,
...
   sycl::event matmul_prim_event = ::dnnl::sycl_interop::execute(
       matmul_prim, handle->get_engine_stream(), matmul_args, transform_events);
 
-  sycl::event scale_d_event;
+  // end of calling oneDNN
+
+  sycl::event scale_d_with_alpha_event;
   if (vector_alpha)
-    scale_d_event = detail::scale_d_with_vector_alpha(
+    scale_d_with_alpha_event = detail::scale_d_with_vector_alpha(
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: If beta == 1, the alpha_vector should be scaled before adding C.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 15]
```diff
@@ -430,18 +557,118 @@ inline void int2float(queue_ptr q_ptr, void *int_ptr, bool is_host_ptr,
...
+            }
+          });
+    });
+  }
+};
 } // namespace detail
 
-/// This function does operation: D = alpha*(A*B) + beta*(C).
+/// This function does operations:
+/// D_temp = epilogue(alpha * scale_a * op(A) * scale_b * op(B) + beta * C)
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: Please be more detailed on formula

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 16]
```diff
@@ -430,18 +557,118 @@ inline void int2float(queue_ptr q_ptr, void *int_ptr, bool is_host_ptr,
...
+/// Aux_d = absmax(D_temp)
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: L661 is not exactly correct, the op depend on setting of matmul_desc_ptr

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 17]
```diff
@@ -673,51 +913,112 @@ inline sycl::event matmul(descriptor_ptr handle, matmul_desc_ptr compute_desc,
...
         q_ptr, m, n, new_d, d_type, alpha, scale_type, {matmul_prim_event});
-  // end of calling oneDNN
+
+  sycl::event amax_d_event;
+  if (auto amax_ptr = compute_desc->_amax_d_pointer) {
+    amax_d_event = detail::type_dispatch<detail::amax_impl>(
+        d_desc->_type, amax_ptr, new_d, new_ldd, d_desc->_rows, d_desc->_cols,
+        q_ptr,
+        std::vector<sycl::event>{matmul_prim_event, scale_d_with_alpha_event});
+    // TODO: Remove this wait. This wait is a WA to make sure the amax is
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: need fix this todo.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp (hunk 14 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp (hunk 17 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: I will remove the amax support now, and will support it later after runtime fix the bug.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 18]
```diff
@@ -128,17 +129,24 @@ class matmul_desc_t {
     trans_a,
     trans_b,
     trans_c,
-    epilogue
+    epilogue,
+    a_scale_pointer,
+    b_scale_pointer,
+    d_scale_pointer,
+    amax_d_pointer,
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: how about use full name for amax in  amax_d_pointer? 
```

Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: done

**Review by zhimingwang36 — APPROVED** [review]

**Review by tomflinda — APPROVED** [review]
LGTM

**Pull request merged**